File: Reaktor/Core/ChemicalState.cpp
// Reaktor is a C++ library for computational reaction modelling.
//
// Copyright (C) 2014 Allan Leal
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program. If not, see <http://www.gnu.org/licenses/>.

#include "ChemicalState.hpp"

// C++ includes
#include <iomanip>
#include <iostream>

// Reaktor includes
#include <Reaktor/Common/Exception.hpp>
#include <Reaktor/Common/Units.hpp>
#include <Reaktor/Core/ChemicalSystem.hpp>
#include <Reaktor/Core/CoreUtils.hpp>

namespace Reaktor {

struct ChemicalState::Impl
{
    /// The chemical system instance
    ChemicalSystem system;

    /// The temperature state of the chemical system (in units of K)
    double T = 298.15;

    /// The pressure state of the chemical system (in units of Pa)
    double P = 1.0e+05;

    /// The molar amounts of the chemical species
    Vector n;

    /// The Lagrange multiplier with respect to the equilibrium charge balance constraint
    double ycharge = 0.0;

    /// The Lagrange multipliers with respect to the equilibrium mass balance constraints (in units of J/mol)
    Vector y;

    /// The Lagrange multipliers with respect to the bound constraints of the species (in units of J/mol)
    Vector z;

    /// Construct a custom ChemicalState::Impl instance
    Impl(const ChemicalSystem& system)
    : system(system)
    {
        n = zeros(system.numSpecies());
        y = zeros(system.numElements());
        z = zeros(system.numSpecies());
    }
};

ChemicalState::ChemicalState(const ChemicalSystem& system)
: pimpl(new Impl(system))
{}

ChemicalState::ChemicalState(const ChemicalState& other)
: pimpl(new Impl(*other.pimpl))
{}

ChemicalState::~ChemicalState()
{}

auto ChemicalState::operator=(ChemicalState other) -> ChemicalState&
{
    pimpl = std::move(other.pimpl);
    return *this;
}

auto ChemicalState::setTemperature(double val) -> void
{
    Assert(val > 0.0, "Cannot set temperature of the chemical state with a non-positive value.", "");
    pimpl->T = val;
}

auto ChemicalState::setTemperature(double val, std::string units) -> void
{
    setTemperature(units::convert(val, units, "kelvin"));
}

auto ChemicalState::setPressure(double val) -> void
{
    Assert(val > 0.0, "Cannot set pressure of the chemical state with a non-positive value.", "");
    pimpl->P = val;
}

auto ChemicalState::setPressure(double val, std::string units) -> void
{
    setPressure(units::convert(val, units, "pascal"));
}

auto ChemicalState::setSpeciesAmounts(double val) -> void
{
    Assert(val >= 0.0,
        "Cannot set the molar amounts of the species.",
        "The given molar abount is negative.");
    pimpl->n.fill(val);
}

auto ChemicalState::setSpeciesAmounts(const Vector& n) -> void
{
    Assert(n.rows() == system().numSpecies(),
        "Cannot set the molar amounts of the species.",
        "The dimension of the molar abundance vector "
        "is different than the number of species.");
    pimpl->n = n;
}

auto ChemicalState::setSpeciesAmount(Index index, double amount) -> void
{
    const unsigned num_species = system().species().size();
    Assert(amount >= 0.0,
        "Cannot set the molar amount of the species.",
        "The given molar amount is negative.");
    Assert(index < num_species,
        "Cannot set the molar amount of the species.",
        "The given index is out-of-range.");
    pimpl->n[index] = amount;
}

auto ChemicalState::setSpeciesAmount(std::string species, double amount) -> void
{
    const Index index = system().indexSpeciesWithError(species);
    setSpeciesAmount(index, amount);
}

auto ChemicalState::setSpeciesAmount(Index index, double amount, std::string units) -> void
{
    setSpeciesAmount(index, units::convert(amount, units, "mol"));
}

auto ChemicalState::setSpeciesAmount(std::string species, double amount, std::string units) -> void
{
    setSpeciesAmount(species, units::convert(amount, units, "mol"));
}

auto ChemicalState::setChargePotential(double ycharge) -> void
{
    pimpl->ycharge = ycharge;
}

auto ChemicalState::setElementPotentials(const Vector& y) -> void
{
    pimpl->y = y;
}

auto ChemicalState::setSpeciesPotentials(const Vector& z) -> void
{
    pimpl->z = z;
}

auto ChemicalState::setVolume(double volume) -> void
{
    Assert(volume >= 0.0, "Cannot set the volume of the chemical state.", "The given volume is negative.");
    const double T = temperature();
    const double P = pressure();
    const Vector& n = speciesAmounts();
    const Vector v = system().phaseVolumes(T, P, n).val;
    const double vtotal = sum(v);
    const double scalar = (vtotal != 0.0) ? volume/vtotal : 0.0;
    scaleSpeciesAmounts(scalar);
}

auto ChemicalState::setPhaseVolume(Index index, double volume) -> void
{
    Assert(volume >= 0.0, "Cannot set the volume of the phase.", "The given volume is negative.");
    Assert(index < system().numPhases(), "Cannot set the volume of the phase.", "The given phase index is out of range.");
    const double T = temperature();
    const double P = pressure();
    const Vector& n = speciesAmounts();
    const Vector v = system().phaseVolumes(T, P, n).val;
    const double scalar = (v[index] != 0.0) ? volume/v[index] : 0.0;
    scaleSpeciesAmountsInPhase(index, scalar);
}

auto ChemicalState::setPhaseVolume(std::string name, double volume) -> void
{
    const Index index = system().indexPhase(name);
    setPhaseVolume(index, volume);
}

auto ChemicalState::scaleSpeciesAmounts(double scalar) -> void
{
    Assert(scalar >= 0.0, "Cannot scale the molar amounts of the species.", "The given scalar is negative.");
    for(unsigned i = 0; i < pimpl->n.rows(); ++i)
        pimpl->n[i] *= scalar;
}

auto ChemicalState::scaleSpeciesAmountsInPhase(Index index, double scalar) -> void
{
    Assert(scalar >= 0.0, "Cannot scale the molar amounts of the species.", "The given scalar `" + std::to_string(scalar) << "` is negative.");
    Assert(index < system().numPhases(), "Cannot set the volume of the phase.", "The given phase index is out of range.");
    const Index start = system().indexFirstSpeciesInPhase(index);
    const Index size = system().numSpeciesInPhase(index);
    for(unsigned i = 0; i < size; ++i)
        pimpl->n[start + i] *= scalar;
}

auto ChemicalState::system() const -> const ChemicalSystem&
{
    return pimpl->system;
}

auto ChemicalState::temperature() const -> double
{
    return pimpl->T;
}

auto ChemicalState::pressure() const -> double
{
    return pimpl->P;
}

auto ChemicalState::speciesAmounts() const -> const Vector&
{
    return pimpl->n;
}

auto ChemicalState::chargePotential() const -> double
{
    return pimpl->ycharge;
}

auto ChemicalState::elementPotentials() const -> const Vector&
{
    return pimpl->y;
}

auto ChemicalState::speciesPotentials() const -> const Vector&
{
    return pimpl->z;
}

auto ChemicalState::speciesAmount(Index index) const -> double
{
    Assert(index < system().numSpecies(),
        "Cannot get the molar amount of the species.",
        "The given index is out-of-range.");
    return pimpl->n[index];
}

auto ChemicalState::speciesAmount(std::string name) const -> double
{
    return speciesAmount(system().indexSpeciesWithError(name));
}

auto ChemicalState::speciesAmount(Index ispecies, std::string units) const -> double
{
    return units::convert(speciesAmount(ispecies), "mol", units);
}

auto ChemicalState::speciesAmount(std::string species, std::string units) const -> double
{
    return units::convert(speciesAmount(species), "mol", units);
}

auto ChemicalState::elementAmounts() const -> Vector
{
    return system().elementAmounts(speciesAmounts());
}

auto ChemicalState::elementAmountsInPhase(Index iphase) const -> Vector
{
    return system().elementAmountsInPhase(iphase, speciesAmounts());
}

auto ChemicalState::elementAmountsInSpecies(const Indices& ispecies) const -> Vector
{
    return system().elementAmountsInSpecies(ispecies, speciesAmounts());
}

auto ChemicalState::elementAmount(Index ielement) const -> double
{
    return system().elementAmount(ielement, speciesAmounts());
}

auto ChemicalState::elementAmount(std::string element) const -> double
{
    return elementAmount(system().indexElementWithError(element));
}

auto ChemicalState::elementAmount(Index index, std::string units) const -> double
{
    return units::convert(elementAmount(index), "mol", units);
}

auto ChemicalState::elementAmount(std::string name, std::string units) const -> double
{
    return units::convert(elementAmount(name), "mol", units);
}

auto ChemicalState::elementAmountInPhase(Index ielement, Index iphase) const -> double
{
    return system().elementAmountInPhase(ielement, iphase, speciesAmounts());
}

auto ChemicalState::elementAmountInPhase(std::string element, std::string phase) const -> double
{
    const unsigned ielement = system().indexElementWithError(element);
    const unsigned iphase = system().indexPhaseWithError(phase);
    return elementAmountInPhase(ielement, iphase);
}

auto ChemicalState::elementAmountInPhase(Index ielement, Index iphase, std::string units) const -> double
{
    return units::convert(elementAmountInPhase(ielement, iphase), "mol", units);
}

auto ChemicalState::elementAmountInPhase(std::string element, std::string phase, std::string units) const -> double
{
    return units::convert(elementAmountInPhase(element, phase), "mol", units);
}

auto ChemicalState::elementAmountInSpecies(Index ielement, const Indices& ispecies) const -> double
{
    return system().elementAmountInSpecies(ielement, ispecies, speciesAmounts());
}

auto ChemicalState::elementAmountInSpecies(Index ielement, const Indices& ispecies, std::string units) const -> double
{
    return units::convert(elementAmountInSpecies(ielement, ispecies), "mol", units);
}

auto operator<<(std::ostream& out, const ChemicalState& state) -> std::ostream&
{
    const ChemicalSystem& system = state.system();
    const double& T = state.temperature();
    const double& P = state.pressure();
    const Vector& n = state.speciesAmounts();
    const Vector u0 = system.standardGibbsEnergies(T, P).val;
    const Vector u  = system.chemicalPotentials(T, P, n).val;
    const Vector ln_a = system.lnActivities(T, P, n).val;
    const Vector a  = ln_a.array().exp();

    out << std::setw(10) << std::left << "Index";
    out << std::setw(20) << std::left << "Species";
    out << std::setw(20) << std::left << "Moles";
    out << std::setw(20) << std::left << "Activity";
    out << std::setw(20) << std::left << "GibbsEnergy";
    out << std::setw(20) << std::left << "ChemicalPotential";
    out << std::endl;
    for(unsigned i = 0; i < system.numSpecies(); ++i)
    {
        out << std::setw(10) << std::left << i;
        out << std::setw(20) << std::left << system.species(i).name();
        out << std::setw(20) << std::left << n[i];
        out << std::setw(20) << std::left << a[i];
        out << std::setw(20) << std::left << u0[i];
        out << std::setw(20) << std::left << u[i];
        out << std::endl;
    }
    return out;
}

auto operator+(const ChemicalState& l, const ChemicalState& r) -> ChemicalState
{
    const Vector& nl = l.speciesAmounts();
    const Vector& nr = r.speciesAmounts();
    ChemicalState res = l;
    res.setSpeciesAmounts(nl + nr);
    return res;
}

auto operator*(double scalar, const ChemicalState& state) -> ChemicalState
{
    ChemicalState res = state;
    res.scaleSpeciesAmounts(scalar);
    return res;
}

auto operator*(const ChemicalState& state, double scalar) -> ChemicalState
{
    return scalar*state;
}

} // namespace Reaktor

File: Reaktoro/Core/ChemicalState.py
# Reaktoro is a unified framework for modeling chemically reactive systems.
#
# Copyright © 2014-2022 Allan Leal
#
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation; either
# version 2.1 of the License, or (at your option) any later version.
#
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
# Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public License
# along with this library. If not, see <http://www.gnu.org/licenses/>.


from reaktoro import *
import numpy as npy
import pytest


def testChemicalState():
    db = SupcrtDatabase("supcrtbl")

    phases = Phases(db)
    phases.add( AqueousPhase("H2O(aq) H+ OH- Na+ Cl- Ca+2 Mg+2 HCO3- CO3-2 CO2(aq) SiO2(aq)") )
    phases.add( GaseousPhase("H2O(g) CO2(g)") )
    phases.add( MineralPhase("Halite") )
    phases.add( MineralPhase("Calcite") )
    phases.add( MineralPhase("Magnesite") )
    phases.add( MineralPhase("Dolomite") )
    phases.add( MineralPhase("Quartz") )

    reactions = Reactions()
    reactions.add( db.reaction("H2O(aq) = H+ + OH-") )
    reactions.add( db.reaction("CO2(aq) + H2O(aq) = H+ + HCO3-") )
    reactions.add( db.reaction("CO2(g) = CO2(aq)") )
    reactions.add( db.reaction("Halite = Na+ + Cl-") )
    reactions.add( db.reaction("Calcite") )

    system = ChemicalSystem(phases, reactions)

    state = ChemicalState(system)

    N = system.species().size()
    n = npy.arange(N)

    assert state.temperature() == 298.15
    assert state.pressure() == 1e5
    assert npy.max(state.speciesAmounts()) == 1e-16  # default initial amount of all species

    state.temperature(30.0, "celsius")
    assert state.temperature() == pytest.approx(303.15)

    state.pressure(1.0, "MPa")
    assert state.pressure() == pytest.approx(1e6)

    state.temperature(40, "celsius")  # testing with int argument
    assert state.temperature() == pytest.approx(313.15)

    state.pressure(2, "MPa")  # testing with int argument
    assert state.pressure() == pytest.approx(2e6)

    state.setTemperature(30.0, "celsius")
    assert state.temperature() == pytest.approx(303.15)

    state.setPressure(1.0, "MPa")
    assert state.pressure() == pytest.approx(1e6)

    state.setTemperature(40, "celsius")  # testing with int argument
    assert state.temperature() == pytest.approx(313.15)

    state.setPressure(2, "MPa")  # testing with int argument
    assert state.pressure() == pytest.approx(2e6)

    state.setSpeciesAmounts(1.0)
    assert npy.min(state.speciesAmounts()) == 1.0
    assert npy.max(state.speciesAmounts()) == 1.0
    assert state.charge() == 1.0 # +1 -1 +1 -1 +2 +2 -1 -2 = 1

    state.setSpeciesAmounts(n)
    assert npy.all(state.speciesAmounts() == n)
    assert state.charge() == -3.0 # +1 -1*2 +1*3 -1*4 +2*5 +2*6 -1*7 -2*8 = -3

    state.setSpeciesAmount(0, 3.1, "mol")
    assert state.speciesAmount(0) == 3.1

    state.setSpeciesAmount(0, 3.1, "mmol")
    assert state.speciesAmount(0).val() == pytest.approx(0.0031)

    state.setSpeciesAmount(N - 1, 4.3, "mol")
    assert state.speciesAmount(N - 1) == 4.3

    state.setSpeciesAmount(N - 1, 4.3, "mmol")
    assert state.speciesAmount(N - 1).val() == pytest.approx(0.0043)

    state.setSpeciesAmount("H2O(aq)", 5.0, "mol")
    assert state.speciesAmount("H2O(aq)") == 5.0

    state.setSpeciesAmount("H2O(aq)", 5.0, "mmol")
    assert state.speciesAmount("H2O(aq)").val() == pytest.approx(0.005)

    state.setSpeciesAmount("Quartz", 10.0, "mol")
    assert state.speciesAmount("Quartz") == 10.0

    state.setSpeciesAmount("Quartz", 10.0, "mmol")
    assert state.speciesAmount("Quartz").val() == pytest.approx(0.01)

    state.setSpeciesMass(0, 7.1, "kg")
    assert state.speciesMass(0) == 7.1

    state.setSpeciesMass(0, 7.1, "mg")
    assert state.speciesMass(0).val() == pytest.approx(7.1e-6)

    state.setSpeciesMass(N - 1, 9.3, "kg")
    assert state.speciesMass(N - 1) == 9.3

    state.setSpeciesMass(N - 1, 9.3, "mg")
    assert state.speciesMass(N - 1).val() == pytest.approx(9.3e-6)

    state.setSpeciesMass("H2O(aq)", 5.0, "kg")
    assert state.speciesMass("H2O(aq)") == 5.0

    state.setSpeciesMass("H2O(aq)", 5.0, "mg")
    assert state.speciesMass("H2O(aq)").val() == pytest.approx(5e-6)

    state.setSpeciesMass("Quartz", 10.0, "kg")
    assert state.speciesMass("Quartz") == 10.0

    state.setSpeciesMass("Quartz", 10.0, "mg")
    assert state.speciesMass("Quartz").val() == pytest.approx(10e-6)

    state.setSpeciesAmount("Calcite", 0.0, "mol")
    state.add("Calcite", 1.0, "mol")
    assert state.speciesAmount("Calcite") == 1.0

    state.add("Calcite", 2.0, "mol")
    assert state.speciesAmount("Calcite") == 3.0

    state.add("Calcite", 5.0, "mmol")
    assert state.speciesAmount("Calcite") == 3.005

    state.setSpeciesAmount("Calcite", 0.0, "mol")
    state.add("Calcite", 2000.0, "g")
    assert state.speciesMass("Calcite") == 2.0  # in kg

    state.add("Calcite", 3000.0, "g")
    assert state.speciesMass("Calcite") == 5.0  # in kg

    state.add("Calcite", 5000.0, "mg")
    assert state.speciesMass("Calcite") == 5.005  # in kg

    state.set("Calcite", 3.0, "mol")
    assert state.speciesAmount("Calcite") == 3.0  # in kg

    state.set("Calcite", 5000.0, "mg")
    assert state.speciesMass("Calcite") == 0.005  # in kg

    scaled = ChemicalState(state)
    scaled.scaleSpeciesAmounts(2.0)
    assert (scaled.speciesAmounts().asarray()
        == state.speciesAmounts().asarray() * 2.0).all()

    scaled = ChemicalState(state)
    scaled.scaleSpeciesAmountsInPhase("AqueousPhase", 3.0)
    assert (scaled.speciesAmountsInPhase("AqueousPhase").asarray()
        == state.speciesAmountsInPhase("AqueousPhase").asarray() * 3.0).all()

    scaled = ChemicalState(state)
    scaled.scaleAmount(1.0, "mol")
    scaled.props().update(scaled)
    assert scaled.props().amount().val() == pytest.approx(1.0)

    scaled = ChemicalState(state)
    scaled.scalePhaseAmount("AqueousPhase", 1.0, "mol")
    scaled.props().update(scaled)
    assert scaled.props().phaseProps("AqueousPhase").amount().val() == pytest.approx(1.0)

    scaled = ChemicalState(state)
    scaled.scalePhaseAmount("GaseousPhase", 1.0, "mol")
    scaled.props().update(scaled)
    assert scaled.props().phaseProps("GaseousPhase").amount().val() == pytest.approx(1.0)

    scaled = ChemicalState(state)
    scaled.scaleFluidAmount(2.0, "mol")
    scaled.props().update(scaled)
    iphases = scaled.props().indicesPhasesWithFluidState()
    assert sum([scaled.props().phaseProps(i).amount().val() for i in iphases]) == pytest.approx(2.0)

    scaled = ChemicalState(state)
    scaled.scaleSolidAmount(3.0, "mol")
    scaled.props().update(scaled)
    iphases = scaled.props().indicesPhasesWithSolidState()
    assert sum([scaled.props().phaseProps(i).amount().val() for i in iphases]) == pytest.approx(3.0)

    scaled = ChemicalState(state)
    scaled.scaleMass(1.0, "kg")
    scaled.props().update(scaled)
    assert scaled.props().mass().val() == pytest.approx(1.0)

    scaled = ChemicalState(state)
    scaled.scalePhaseMass("AqueousPhase", 1.0, "kg")
    scaled.props().update(scaled)
    assert scaled.props().phaseProps("AqueousPhase").mass().val() == pytest.approx(1.0)

    scaled = ChemicalState(state)
    scaled.scaleFluidMass(2.0, "kg")
    scaled.props().update(scaled)
    iphases = scaled.props().indicesPhasesWithFluidState()
    assert sum([scaled.props().phaseProps(i).mass() for i in iphases]) == pytest.approx(2.0)

    scaled = ChemicalState(state)
    scaled.scaleSolidMass(3.0, "kg")
    scaled.props().update(scaled)
    iphases = scaled.props().indicesPhasesWithSolidState()
    assert sum([scaled.props().phaseProps(i).mass() for i in iphases]) == pytest.approx(3.0)

    scaled = ChemicalState(state)
    scaled.scaleVolume(1.0, "m3")
    scaled.props().update(scaled)
    assert scaled.props().volume().val() == pytest.approx(1.0)

    scaled = ChemicalState(state)
    scaled.scalePhaseVolume("AqueousPhase", 1.0, "m3")
    scaled.props().update(scaled)
    assert scaled.props().phaseProps("AqueousPhase").volume().val() == pytest.approx(1.0)

    scaled = ChemicalState(state)
    scaled.scalePhaseVolume("GaseousPhase", 1.0, "m3")
    scaled.props().update(scaled)
    assert scaled.props().phaseProps("GaseousPhase").volume().val() == pytest.approx(1.0)

    scaled = ChemicalState(state)
    scaled.scaleFluidVolume(2.0, "m3")
    scaled.props().update(scaled)
    iphases = scaled.props().indicesPhasesWithFluidState()
    assert sum([scaled.props().phaseProps(i).volume().val() for i in iphases]) == pytest.approx(2.0)

    scaled = ChemicalState(state)
    scaled.scaleSolidVolume(3.0, "m3")
    scaled.props().update(scaled)
    iphases = scaled.props().indicesPhasesWithSolidState()
    assert sum([scaled.props().phaseProps(i).volume().val() for i in iphases]) == pytest.approx(3.0)

    props = state.props()
    props.update(state)

    assert props.temperature() == state.temperature()
    assert props.pressure() == state.pressure()
    assert props.speciesAmounts() == state.speciesAmounts()
    assert state.charge() == -3.0

    #-------------------------------------------------------------------------
    # TESTING METHOD: ChemicalState.update
    #-------------------------------------------------------------------------

    otherstate = ChemicalState(system)

    otherstate.update(state.temperature(), state.pressure(), state.speciesAmounts())

    assert otherstate.temperature() == state.temperature()
    assert otherstate.pressure() == state.pressure()
    assert otherstate.speciesAmounts() == state.speciesAmounts()

    state.props().update(state)

    assert otherstate.props().temperature() == state.temperature()
    assert otherstate.props().pressure() == state.pressure()
    assert otherstate.props().speciesAmounts() == state.speciesAmounts()
    assert otherstate.props().speciesActivitiesLn() == state.props().speciesActivitiesLn()

    #-------------------------------------------------------------------------
    # TESTING METHOD: ChemicalState.update
    #-------------------------------------------------------------------------

    otherstate.updateIdeal(state.temperature(), state.pressure(), state.speciesAmounts())

    assert otherstate.temperature() == state.temperature()
    assert otherstate.pressure() == state.pressure()
    assert otherstate.speciesAmounts() == state.speciesAmounts()

    state.props().updateIdeal(state)

    assert otherstate.props().temperature() == state.temperature()
    assert otherstate.props().pressure() == state.pressure()
    assert otherstate.props().speciesAmounts() == state.speciesAmounts()
    assert otherstate.props().speciesActivitiesLn() == state.props().speciesActivitiesLn()

    #-------------------------------------------------------------------------
    # TESTING METHOD: ChemicalState.assign
    #-------------------------------------------------------------------------
    state = ChemicalState(system)
    otherstate = ChemicalState(system)

    state.setTemperature(321, "K")
    state.setPressure(432, "MPa")
    state.setSpeciesAmounts(1.23)

    otherstate.assign(state)

    assert otherstate.temperature() == state.temperature()
    assert otherstate.pressure() == state.pressure()
    assert otherstate.speciesAmounts() == state.speciesAmounts()

    #-------------------------------------------------------------------------
    # TESTING METHOD: ChemicalState.clone
    #-------------------------------------------------------------------------
    state = ChemicalState(system)
    state.setTemperature(321, "K")

    otherstate = state.clone()

    assert otherstate.temperature() == state.temperature()

    otherstate.setTemperature(1234, "K")

    assert state.temperature() == 321  # ensure state was not changed with changing otherstate above!
